### PR TITLE
HipChat Support

### LINKED ISSUE #968: HipChat support
We need to re-implement HipChat support in terms of the new chat provider API.

### PR DESCRIPTION
**Pull request by kevsmith**
This PR reinstates HipChat support on top of the new chat provider API.

- Supports DMs, rooms, and private rooms
- Supports both @ mentions and spoken commands
- Full support for Greenbar templates
- Replaced `AdapterCase` and provider-specific helper modules with `ProviderCase` and provider-specific stateful testing client to simplify integration testing. Tests will wait for an appropriate reply from Cog instead of polling chat history.

Fixes #968

### CODE REVIEW

**Comment by pjmorr** [comment]
+1

**Comment by kevsmith** [comment]
Thanks to @pjmorr for testing HipChat support w/his private HipChat server

File: lib/cog/chat/hipchat/connector.ex [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                     notify: false,
+                     message: text})
+  end
+
+  defp send_room_message(room, message, state) do
+    body = prepare_message(message)
+    url = Enum.join([state.api_root, "room", room.secondary_id, "notification"], "/") <> "?token=#{state.api_token}"
+    response = HTTPotion.post(url, headers: ["Content-Type": "application/json",
+                                             "Accepts": "application/json",
+                                             "Authorization": "Bearer #{state.api_token}"],
```
Comment (vanstee) [inline lib/cog/chat/hipchat/connector.ex]: Do we need to set the token as a query param and a header? That's kinda weird.

File: lib/cog/chat/hipchat/connector.ex [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      body: body)
```
Comment (vanstee) [inline lib/cog/chat/hipchat/connector.ex]: Might also be nice to wrap up this request stuff into a helper function. Totally not necessary though.

File: lib/cog/repository/chat_providers.ex [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  @type error :: Ecto.Changeset.error | atom
+
+  alias Cog.Repo
+  alias Cog.Models.ChatProvider
+
+  @spec get_provider_state(binary()) :: nil | map()
+  def get_provider_state(name, default \\ %{}) when is_binary(name) do
+    case Repo.get_by!(ChatProvider, name: name) do
+      nil ->
+        nil
```
Comment (vanstee) [inline lib/cog/repository/chat_providers.ex]: IIRC `Repo.get_by!` can't ever return `nil`.

File: config/dev.exs [hunk 4]
```diff
@@ -13,9 +13,19 @@ config :cog,
   :template_cache_ttl, {1, :sec}
 
 config :cog, Cog.Chat.Adapter,
-  providers: [slack: Cog.Chat.Slack.Provider],
+  providers: [#slack: Cog.Chat.Slack.Provider,
+              hipchat: Cog.Chat.HipChat.Provider],
```
Comment (kevsmith) [inline config/dev.exs]: Ah yeah. Joys of poking at `$NEW_FEATURE` locally.....

Will fix.

**Pull request merged**